backend/de.metas.handlingunits.base/src/main/java/org/adempiere/mm/attributes/spi/impl/HUVendorBPartnerAttributeValuesProvider.java

```diff
@@ -0,0 +1,302 @@
+
+	static final String ATTRIBUTEVALUETYPE = X_M_Attribute.ATTRIBUTEVALUETYPE_Number;
+	private static final String CACHE_PREFIX = IAttributeDAO.CACHEKEY_ATTRIBUTE_VALUE;
+
+	private static final CtxName CTXNAME_M_HU_ID = CtxNames.parse("M_HU_ID/-1");
+	private static final CtxName CTXNAME_CurrentVendor_BPartner_ID = CtxNames.parse("CurrentVendor_BPartner_ID/-1");
+
+
+	private static final CCache<Integer, List<KeyNamePair>> vendors =  CCache.<Integer, List<KeyNamePair>> builder()
+			.tableName(IAttributeDAO.CACHEKEY_ATTRIBUTE_VALUE)
+			.cacheName(CACHE_PREFIX)
+			.initialCapacity(20)
+			.build();
+
+	private final CCache<ArrayKey, List<KeyNamePair>> hu2Vendors = CCache.<ArrayKey, List<KeyNamePair>> builder()
+			.cacheName(vendors.getCacheName() + "#AndHU")
+			.initialCapacity(100)
+			.expireMinutes(10)
+			.build();
+
+	private final I_M_Attribute attribute;
+
+	public HUVendorBPartnerAttributeValuesProvider(@NonNull final I_M_Attribute attribute)
+	{
+		this.attribute = attribute;
+	}
+
+	@Override
+	public String getCachePrefix()
+	{
+		return CACHE_PREFIX;
+	}
+
+	@Override
+	public List<CCacheStats> getCacheStats()
+	{
+		return ImmutableList.of(vendors.stats(), hu2Vendors.stats());
+	}
+
+	@Override
+	public String getAttributeValueType()
+	{
+		return ATTRIBUTEVALUETYPE;
+	}
+
+	@Override
+	public final KeyNamePair getNullValue()
+	{
+		return staticNullValue();
+	}
+
+	static KeyNamePair staticNullValue()
+	{
+
+		final ITranslatableString displayNameTrl = iMsgBL.translatable("NoneOrEmpty");
+
+		final String adLanguage = Env.getAD_Language(Env.getCtx());
+		final String displayName = adLanguage != null ? displayNameTrl.translate(adLanguage) : displayNameTrl.getDefaultValue();
+
+		// NOTE: we use KeyNamePair's Key=0 because "-1" is specially handled by KeyNamePair (see KeyNamePair.getID() which returns null)
+		// and we run in some weird problems
+		return KeyNamePair.of(0, displayName);
+
+	}
+
+	@Override
+	public boolean isHighVolume()
+	{
+		return true;
+	}
+
+	@Override
+	public boolean isAllowAnyValue()
+	{
+		return true;
+	}
+
+	@Override
+	public Evaluatee prepareContext(final IAttributeSet attributeSet)
+	{
+		final I_M_HU hu = huAttributesBL.getM_HU(attributeSet);
+		final int bpartnerId = hu.getC_BPartner_ID();
+		final int huId = hu.getM_HU_ID();
+
+		return Evaluatees.mapBuilder()
+				.put(CTXNAME_M_HU_ID, huId > 0 ? huId : -1)
+				.put(CTXNAME_CurrentVendor_BPartner_ID, bpartnerId > 0 ? bpartnerId : -1)
+				.build();
+	}
+
+	@Override
+	public List<? extends NamePair> getAvailableValues(final Evaluatee evalCtx)
+	{
+		final int huId = CTXNAME_M_HU_ID.getValueAsInteger(evalCtx);
+		final int currentVendorId = CTXNAME_CurrentVendor_BPartner_ID.getValueAsInteger(evalCtx);
+
+		//
+		// Create the cache key based on M_HU_ID and current vendor
+		final ArrayKey cacheKey = ArrayKey.of(
+				huId <= 0 ? System.currentTimeMillis() : huId,      // if M_HU_ID <= 0 then make it unique
```

Comment:
instead of making it unique.... it better don't cache it at all.
don't have to pollute the cache if is not needed.